rust/headless-client/src/windows.rs

```diff
@@ -0,0 +1,489 @@
+                }
+                Poll::Pending => {}
+            }
+
+            match ipc_service.as_mut().poll(cx) {
+                Poll::Ready(Ok(())) => {
+                    return Poll::Ready(Err(anyhow::anyhow!(
+                        "Impossible, ipc_listencan't return Ok"
+                    )));
+                }
+                Poll::Ready(Err(error)) => {
+                    return Poll::Ready(Err(error).context("ipc_listen failed"));
+                }
+                Poll::Pending => {}
+            }
+
+            Poll::Pending
+        })
+        .await
+    })
+}
+
+// Generates `ffi_service_run` from `service_run`
+windows_service::define_windows_service!(ffi_service_run, windows_service_run);
+
+fn windows_service_run(_arguments: Vec<OsString>) {
+    if let Err(_e) = fallible_windows_service_run() {
+        todo!();
+    }
+}
+
+#[cfg(debug_assertions)]
+const SERVICE_RUST_LOG: &str = "debug";
+
+#[cfg(not(debug_assertions))]
+const SERVICE_RUST_LOG: &str = "info";
+
+// Most of the Windows-specific service stuff should go here
+fn fallible_windows_service_run() -> Result<()> {
+    let log_path =
+        crate::known_dirs::ipc_service_logs().context("Can't compute IPC service logs dir")?;
+    std::fs::create_dir_all(&log_path)?;
+    let (layer, _handle) = file_logger::layer(&log_path);
+    let filter = EnvFilter::from_str(SERVICE_RUST_LOG)?;
+    let subscriber = Registry::default().with(layer.with_filter(filter));
+    set_global_default(subscriber)?;
+    tracing::info!(git_version = crate::GIT_VERSION);
+
+    let rt = tokio::runtime::Runtime::new()?;
+    let (shutdown_tx, mut shutdown_rx) = mpsc::channel(1);
+
+    let event_handler = move |control_event| -> ServiceControlHandlerResult {
+        tracing::debug!(?control_event);
+        match control_event {
+            // TODO
+            ServiceControl::Interrogate => ServiceControlHandlerResult::NoError,
+            ServiceControl::Stop => {
+                tracing::info!("Got stop signal from service controller");
+                shutdown_tx.blocking_send(()).unwrap();
+                ServiceControlHandlerResult::NoError
+            }
+            ServiceControl::UserEvent(_) => ServiceControlHandlerResult::NoError,
+            ServiceControl::Continue
+            | ServiceControl::NetBindAdd
+            | ServiceControl::NetBindDisable
+            | ServiceControl::NetBindEnable
+            | ServiceControl::NetBindRemove
+            | ServiceControl::ParamChange
+            | ServiceControl::Pause
+            | ServiceControl::Preshutdown
+            | ServiceControl::Shutdown
+            | ServiceControl::HardwareProfileChange(_)
+            | ServiceControl::PowerEvent(_)
+            | ServiceControl::SessionChange(_)
+            | ServiceControl::TimeChange
+            | ServiceControl::TriggerEvent => ServiceControlHandlerResult::NotImplemented,
+            _ => ServiceControlHandlerResult::NotImplemented,
```

Comment:
Interesting, are these notifications that Windows will send us? Anything useful we can hook for `reconnect` / `set_dns`?

Reply:
Yeah the Windows service controller will send these. It might be useful but I'm not sure what they all mean